Technotes/001-async-networking.md

```diff
@@ -0,0 +1,246 @@
+
+```swift
+public protocol ImageProcessing {
+    func decompress(_ data: Data) async throws -> Image
+    func process(_ data: Data, configuration: ImageConfiguration) async throws -> Image
+}
+```
+
+Until custom executors are supported it seems that Swift Concurrency doesn't really support any attempt to limit the number of threads dedicated to a particular task / set of tasks. The main idea behind this separate object is to abstract this usage of the older concurrency model.
+
+I do expect (and preliminary testing seems to indicate) the cooperative thread pool to be good at handling network requests and cooperative task cancellation.
+
+## Notable Changes
+
+### The Collection of Active and Pending Tasks
+
+This proposal includes a change to the collection of active / pending tasks used, in part, to support the `ImageFetcher.cancel(_:)` method that deserves additional discussion:
+
+- Current: `var tasks: Set<ImageFetcherTask>`
+- Proposed: `var tasks: Dictionary<String, Task<ImageSource, Error>>`
+
+#### Current `tasks` Implementation
+
+In the current implementation each call to `ImageFetcher.task(_:)` and `.load(_:)` creates a new task if the requested resource is not cached. That means that if two different sections of code request the same resource and one subsequently cancels its request, the other is unaffected. However, that also means the caller is responsible for preventing duplicate requests in cases where that is likely to be a concern.
+
+The `ImageFetcherTask` provides access to the `configuration`, `handler`, and `result` of a given operation. While it seems like it should be very unlikely, this present a potential thread-safety issue as it is possible for the caller to try reading a task’s `result` on one thread while the underlying `Operation`’s `completionBlock` mutates the task’s `result` on another thread or for the caller to mutate its `configuration` or `handler` member while it is being read on another thread.
+
+#### Proposed `tasks` Implementation
+
+This proposal changes the `tasks` collection to a `Dictionary` and first checks for an active/pending for the requested configuration before checking for a cached image or requesting one. This absolves the caller of responsibility for preventing duplicate requests, but it also means that any call to cancel an active / pending request cancels it for all callers.
+
+I replaced `ImageFetcherTask` by providing direct access to the underlying `Task`, which offers access to the request’s eventual value and a `cancel()` method. The result is more restricted but safer since we remove possibility of concurrently reading/writing task members.
+
+`ImageFetcher` will use an `NSLock` to ensure thread-safe access to `tasks`. I considered making it an `actor` but that seems like overkill when we only need to protect `tasks` and all access to it will be very quick.
```

Comment:
Given that there is no guarantee a task will resume on the same thread, I assume locks will not be held across `await`s?

Reply:
Right. I was thinking all access to `tasks` would have to be through the following members: 

```swift
// MARK: Task Access
private extension ImageFetcher {
    func insertTask(_ imageFetcherTask: Task<ImageSource, Error>, key: ImageConfiguration) {
        lock.lock()
        defer { lock.unlock() }

        tasks[key.key] = imageFetcherTask
    }

    func getTask(_ key: ImageConfiguration) -> Task<ImageSource, Error>? {
        lock.lock()
        defer { lock.unlock() }

        return tasks[key.key]
    }

    @discardableResult
    func removeTask(_ key: ImageConfiguration) -> Task<ImageSource, Error>? {
        lock.lock()
        defer { lock.unlock() }

        return tasks.removeValue(forKey: key.key)
    }
}
```

Unlike with an `Actor`, we don't have the advantage of having the compiler warn us if we decide to access `tasks` outside of those methods, but this will be much more performant -- by more than an [order of magnitude](https://trycombine.com/posts/performance-actor-queue-lock-benchmark/) by some simple benchmarks -- and will allow us to avoid a bunch of `nonisolated` for all the methods that don't involve access to `tasks`.